File: Archived/14단계/12865.py
import sys

N, K = map(int,sys.stdin.readline().strip().split())
li = []
for i in range(N):
    w,v = map(int,sys.stdin.readline().strip().split())
    li.append((w,v))
cache = [[0 for i in range(K+1)] for j in range(N+1)]

for n in range(1,N+1):
    w,v = li[n-1][0], li[n-1][1]
    for k in range(1,K+1):
        if w > k:
            cache[n][k] = cache[n-1][k]
            continue
        
        cache[n][k] = max(cache[n-1][k], v + cache[n-1][k-w])
print(cache[N][K])

File: Archived/DP/12865.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int N,K;
    static int[][] values;
    static int[] cache;
    public static void main(String[] args) throws IOException{
        int[] tmp = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        N = tmp[0]; K = tmp[1];
        values = new int[N][];
        cache = new int[K+1];
        for(int i=0;i<N;i++) 
            values[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int ans=0;
        for(int i=0;i<N;i++){
            for(int j=K;j>=0;j--){
                if(j-values[i][0]>=0)
                    cache[j] = Math.max(cache[j], cache[j-values[i][0]]+values[i][1]);
            }
        }
        ans = Arrays.stream(cache).max().getAsInt();
        System.out.println(ans);
    }
}